### DISCUSSION TITLE
ComputeBuffer.Setdata

### DISCUSSION THREAD

**Discussion by Deytooh** [Tencent/puerts#850] [General]
[Unity] Hey guys, I want to use SetData from Computebuffer class but I keep getting the invalid arguments error. 
I did Object.setPrototypeOf(ComputeBuffer,ComputeBuffer.prototype) but still it doesnt work :(

**Comment by zombieyang** [comment]
could you provide some of your code where the error was thrown?

**Reply by Deytooh** [reply]
```
private args: int[] = [0, 0, 0, 0, 0];

Start()
{
    this.argsBuffer = new UnityEngine.ComputeBuffer(1, this.args.length * 4, ComputeBufferType.IndirectArguments);
    this.UpdateBuffer();
}

// Indirect args
UpdateBuffer()
{

        if (this.instanceMesh != null) {
            this.args[0] = this.instanceMesh.GetIndexCount(this.subMeshIndex);
            this.args[1] = this.instanceCount;
            this.args[2] = this.instanceMesh.GetIndexStart(this.subMeshIndex);
            this.args[3] = this.instanceMesh.GetBaseVertex(this.subMeshIndex);
        }
        else {
            this.args[0] = this.args[1] = this.args[2] = this.args[3] = 0;
        }
    
        this.argsBuffer.SetData(this.args);
}
```

**Comment by zombieyang** [comment]
`private args: int[] = [0, 0, 0, 0, 0];` is a js array. the `SetData` api needs a csharp array or list.

you can refer to this:
https://github.com/chexiongsheng/puerts_unity_demo/blob/master/projects/0_Basic_Demo/TsProj/QuickStart.ts#L68
https://github.com/chexiongsheng/puerts_unity_demo/blob/master/projects/0_Basic_Demo/Assets/Examples/05_Typescript/SomeClasses.cs#L101

PuerTS has no plan to support an implicit translation for this. but you can write a util function to do this.

**Reply by Deytooh** [reply]
Thank you for your answer ! :) so I tried to change to Uint32Array type but it still throw me an error sadly

**Reply by zombieyang** [reply]
A JS `Uint32Array` will become a `Puerts.ArrayBuffer` in CSharp :).
Have you tried to change to a `csharp.System.Collections.Generic.List$1 `?

**Reply by Deytooh** [reply]
Ohh I see ! 👍  But yeah I already tried with the List$1 but still does not work 😢

**Reply by zombieyang** [reply]
I wrote a piece of code that works:
```
        const CS = require('csharp');
        const puerts = require('puerts');
        const UIntList = puerts.$generic(CS.System.Collections.Generic.List$1, CS.System.UInt32)

        var args = new UIntList();
        var argsBuffer = new CS.UnityEngine.ComputeBuffer(1, 20, 256);
        var mf = this.bindTo.GetComponent(puerts.$typeof(CS.UnityEngine.MeshFilter));
        args.Add(mf.mesh.GetIndexCount(0));
        args.Add(0);
        args.Add(mf.mesh.GetIndexStart(0));
        args.Add(mf.mesh.GetBaseVertex(0));
        args.Add(0);
        argsBuffer.SetData(args.ToArray());
        
        argsBuffer.Release();
```
Seems like there are two problems: 
1. cannot get the correct value of `ComputeBufferType.IndirectArguments` is JS. So I just write 256. Maybe it is a bug. If you confirm that you get an invalid value by invoking `console.log(CS.ComputeBufferType.IndirectArguments)`, you can write an issue for that and I will fix it later.
2. only `SetData(System.Array)` is worked in JS now. `SetData(List<uint>)` is a generic function so is not valid in this version. Maybe the next version can make `SetData(List<uint>)` available: #819 .

**Reply by Deytooh** [reply]
Oh great ! thank you very much for your help ! I have no errors anymore on the SetData :) great ! 
Actually I need to use this buffer with Graphics.DrawMeshInstancedIndirect method but it throws me an error also ( sorry xD )

`        Graphics.DrawMeshInstancedIndirect(this.instanceMesh, 0, this.instanceMaterial, new Bounds(Vector3.zero, new Vector3(100.0, 100.0, 100.0)), this.argsBuffer);
`

I get this error -> Error: c# exception:Index was outside the bounds of the array.,stack:  at Puerts.Parameters.FillByRefParameters (Puerts.CallInfo callInfo)

**Reply by Deytooh** [reply]
Okay I just needed to fill every arguments even if they are not mandatory ! no error anymore ~ thank you :)
